src/mon/OSDMonitor.cc

```diff
@@ -1481,6 +1482,21 @@ bool OSDMonitor::should_propose(double& delay)
     return true;
   }
 
+  // propose as fast as possible if updating up_thru or pg_temp
+  // want to merge OSDMap changes as much as possible
+  if ((pending_inc.new_primary_temp.size() == 1
+      || pending_inc.new_up_thru.size() == 1)
+      && pending_inc.new_state.size() < 2) {
+    dout(15) << " propose as fast as possible for up_thru/pg_temp" << dendl;
+
+    utime_t now = ceph_clock_now(NULL);
+    if (now - last_no_wait_time > g_conf->paxos_propose_interval) {
+      delay = 0.0;
+      last_no_wait_time = now;
+      return true;
+    }
+  }
```

Comment:
This logic doubles the number of OSDMaps we'd otherwise generate — we'll do two every paxos_propose_interval. You'd need something at least a little more sophisticated. Maybe it would work if you renamed last_no_wait_time -> last_attempted_nowait_time, set it every time you run this check, and only allow it to succeed every *two* paxos_propose_intervals? Or just only let it go through if at least a paxos_propose_interval has passed since the previous OSDMap was generated?